Store the original "CREATE VIEW" statement in the Babelfish catalog

shalinilohia50 opened the pull request:
Currently, Babelfish stores the modified DDL for "CREATE VIEW" statement not the original DDL used by the customer to create that view.

Task: BABEL-3835
Signed-off-by: Shalini Lohia <lshalini@amazon.com>

### Description

[Describe what this change achieves - Guidelines below (please delete the guidelines after writing the PR description)]

> 1. *What* is the change? This is best described in terms of “Currently, Babelfish does X. With this change it now does Y.” Think of “What *did* it *used* to do?” and “What *does* it do *now*?”
2. *Why* was the change made? What drove our desire to put effort into the change?
3. *How* was the code changed should only appear for large commits. This can serve as a rough roadmap to what’s contained in the commit. It should be very high level; if it’s directly referencing code it’s probably too detailed. It’s also critical that this section of a commit message does not try to replace proper code documentation (ie, block comments or README files). Generally, this section should only appear if the commit itself is large enough that it’s helpful to provide a roadmap to someone looking at the commit.
4. The last descriptive piece is the “title” for the commit: the very first line of the commit message, which should typically be less than 80 characters. A good title is *critical*, because it’s the only thing that shows up in places like the Github commit listing. No one’s got time to read through full commit messages when trying to find a single commit out of dozens.


### Issues Resolved

[List any issues this PR will resolve]

### Test Scenarios Covered ###
* **Use case based -**


* **Boundary conditions -**


* **Arbitrary inputs -**


* **Negative test cases -**


* **Minor version upgrade tests -**


* **Major version upgrade tests -**


* **Performance tests -**


* **Tooling impact -**


* **Client tests -**

lohia-shalini committed:
Store the original "CREATE VIEW" statement in the Babelfish catalog

Currently, Babelfish stores the modified DDL for "CREATE VIEW" statement
not the original DDL used by the customer to create that view.

Task: BABEL-3835
Signed-off-by: Shalini Lohia <lshalini@amazon.com>

lohia-shalini committed:
Handle the case for sp_describe_first_result_set

lohia-shalini committed:
Add a test

lohia-shalini committed:
Add tests to show the original stmt

lohia-shalini committed:
Fix the upgrade test

lohia-shalini committed:
Empty-Commit

lohia-shalini committed:
Merge branch 'BABEL_2_X_DEV' into jira-babel-3835

lohia-shalini committed:
Update test output with upgrade

lohia-shalini committed:
Fix upgrade test for 14_5

lohia-shalini committed:
Remove a test from 14_6 schedule

shalinilohia50 committed:
Merge branch 'BABEL_2_X_DEV' into jira-babel-3835

2jungkook commented on contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--2.3.0--2.4.0.sql:
```diff
@@ -726,6 +726,9 @@ WHERE has_schema_privilege(s.schema_id, 'USAGE')
 GRANT SELECT ON sys.check_constraints TO PUBLIC;
 
+-- For all the views created on previous versions, the definition in the catalog should be NULL.
+UPDATE sys.babelfish_view_def SET definition = NULL;
```
As 3_X_DEV is the default branch now, could you raise the PR for 3_X_DEV first?

Also please be careful that there are two different upgrade paths:
1. 2.3 --> 2.4 --> 3.0 --> 3.1
2. 2.3 --> 3.0 --> 3.1

shalinilohia50 replied:
#1046 https://github.com/babelfish-for-postgresql/babelfish_extensions/pull/1046

2jungkook commented on contrib/babelfishpg_tsql/src/hooks.c:
```diff
@@ -1922,7 +1923,14 @@ pltsql_store_view_definition(const char *queryString, ObjectAddress address)
+	 * we don't want to save the view definition for that case.
+	 */
+	else
+		new_record[3] = CStringGetTextDatum("NULL");
```
1. This will store a "NULL" string, not a NULL value. Is this intended?
2. Do we have a test case for this?

shalinilohia50 replied:
In case the original_query is NULL, we have inserted "queryString" in the catalog.
This executes for a scenario where we internally translate "sp_describe_first_result_set" to a "CREATE VIEW" statement, though we don't save anything in the catalog since it's not an actual view. 
IMO we can't have a test case for this.

2jungkook commented on contrib/babelfishpg_tsql/src/tsqlIface.cpp:
```diff
@@ -2905,6 +2905,7 @@ handleBatchLevelStatement(TSqlParser::Batch_level_statementContext *ctx, tsqlSel
 	result->body = list_make1(init);
 	// create PLtsql_stmt_execsql to wrap all query string
 	PLtsql_stmt_execsql *execsql = (PLtsql_stmt_execsql *) makeSQL(ctx);
+	execsql->original_query = (makeTsqlExpr(ctx, false))->query;
```
It seems that we are passing `char *` from here to `pltsql_store_view_definition` hook. Could you double check the life cycle of this pointer? In other words, `(makeTsqlExpr(ctx, false))->query` is in which memory context?

shalinilohia50 replied:
Sorry I did not understand what's your question here. 
At this point, _ctx_ is the batch_level_statement context and it contains the original statement throughout. I am making use of that _ctx_ to get the query and save it in _execsql->original_query_

2jungkook commented on contrib/babelfishpg_tsql/src/pl_exec.c:
```diff
@@ -10207,3 +10211,21 @@ bool reset_search_path(PLtsql_stmt_execsql *stmt, char *old_search_path, bool* r
+ * Set the global variable original_query_string with the original query.
+ */
+static void
+set_original_query_string(char *queryString)
```
Is there a specific reason to have this function?

shalinilohia50 replied:
It's not needed anymore. Removed it.

2jungkook commented on contrib/babelfishpg_tsql/src/pl_exec.c:
```diff
@@ -142,6 +142,7 @@ static SimpleEcontextStackEntry *simple_econtext_stack = NULL;
 	MemoryContextAlloc(get_eval_mcontext(estate), sz)
 #define eval_mcontext_alloc0(estate, sz) \
 	MemoryContextAllocZero(get_eval_mcontext(estate), sz)
+char *original_query_string;
```
Let's initialize this to NULL. Otherwise, it could have a garbage value and `if (original_query)` may succeed.

shalinilohia50 replied:
Initialized "original_query" to NULL.

2jungkook commented on test/JDBC/upgrade/14_6/schedule:
```diff
@@ -252,7 +252,6 @@ format
 msdb-dbo-syspolicy_system_health_state
 dateadd_internal_df
 sys-all_columns
-sys-all_sql_modules
```
Why did you remove this?

shalinilohia50 replied:
This view used the catalog table babelfish_view_def for the view definition which has been modified, so we have to write separate test files in **sys-all_sql_modules-before-14_7** format. Until then, removing this test.

2jungkook commented:
1. Please fill in PR description properly, e.g., https://github.com/babelfish-for-postgresql/babelfish_extensions/pull/911)
2. Instead of using the preparation folder, please follow this trick (https://github.com/babelfish-for-postgresql/babelfish_extensions/pull/270/files), i.e.,
```
- bbf_view_definition-before-14-7-vu-prepare/verify/cleanup
- bbf_view_definition-vu-prepare/verify/cleanup
- add those to proper schedule files including latest
```

shalinilohia50 commented:
> 1. Please fill in PR description properly, e.g., [Remove GetConfigOption call from TDSSendDone function #911](https://github.com/babelfish-for-postgresql/babelfish_extensions/pull/911))
> 2. Instead of using the preparation folder, please follow this trick (https://github.com/babelfish-for-postgresql/babelfish_extensions/pull/270/files), i.e.,
> 
> ```
> - bbf_view_definition-before-14-7-vu-prepare/verify/cleanup
> - bbf_view_definition-vu-prepare/verify/cleanup
> - add those to proper schedule files including latest
> ```

Updated the PR description.
Updated the tests for bbf_view_def, there were redundant test files.

lohia-shalini committed:
Merge branch 'BABEL_2_X_DEV' into jira-babel-3835

lohia-shalini committed:
Update the code
